src/compiler/checker.ts

```diff
@@ -50901,10 +50901,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
     }
 
-    function findBestTypeForObjectLiteral(source: Type, unionTarget: UnionOrIntersectionType) {
-        if (getObjectFlags(source) & ObjectFlags.ObjectLiteral && someType(unionTarget, isArrayLikeType)) {
-            return find(unionTarget.types, t => !isArrayLikeType(t));

```

Comment:
The referenced bug was a somewhat funny one because optional properties always include `| undefined` in their types and `undefinedType` is always kept at the beginning of `unionType.types`. So in such a situation this was **always** returning that `undefinedType` - making it impossible to create a good elementwise elaboration.